src/target/riscv/riscv-013_reg.c

```diff
@@ -13,73 +13,452 @@
+	riscv_reg_t value;
+	int res = riscv013_get_register(riscv_reg_impl_get_target(reg),
+			&value, reg->number);
+	if (res != ERROR_OK)
+		return res;
+	buf_set_u64(reg->value, 0, reg->size, value);
+	return res;
+}
 
-	/* TODO: Hack to deal with gdb that thinks these registers still exist. */
-	if (reg->number > GDB_REGNO_XPR15 && reg->number <= GDB_REGNO_XPR31 &&
-			riscv_supports_extension(target, 'E')) {
-		buf_set_u64(reg->value, 0, reg->size, 0);
-		return ERROR_OK;
-	}
+static int send_riscv013_reg(const struct reg *reg, const uint8_t *buf)
+{
+	assert(reg->size <= 64);
+	const riscv_reg_t value = buf_get_u64(buf, 0, reg->size);
+	return riscv013_set_register(riscv_reg_impl_get_target(reg),
+			reg->number, value);
+}
 
-	if (reg->number >= GDB_REGNO_V0 && reg->number <= GDB_REGNO_V31) {
-		if (riscv013_get_register_buf(target, reg->value, reg->number) != ERROR_OK)
-			return ERROR_FAIL;
+/* TODO: Inline "riscv013_*_register_buf" into "vreg" accessors.
+ */
+static int fetch_vreg(struct reg *reg)
+{
+	return riscv013_get_register_buf(riscv_reg_impl_get_target(reg),
+			reg->value, reg->number);
+}
 
-		reg->valid = riscv_reg_impl_gdb_regno_cacheable(reg->number, /* is write? */ false);
-	} else {
-		uint64_t value;
-		int result = riscv_reg_get(target, &value, reg->number);
-		if (result != ERROR_OK)
-			return result;
-		buf_set_u64(reg->value, 0, reg->size, value);
+static int send_vreg(const struct reg *reg, const uint8_t *buf)
+{
+	return riscv013_set_register_buf(riscv_reg_impl_get_target(reg),
+			reg->number, reg->value);
+}
+
+/* End of register fetch/send definitions. */
+/* Start of cache entry utils. */
+
+static void set_cache_value(struct reg *reg, const uint8_t *buf)
+{
+	assert(reg);
+	assert(reg->value);
+	assert(reg->size);
+	buf_cpy(buf, reg->value, reg->size);
+}
+
+/**
+ * Note: there are only 3 states a cahce entry can be in:
+ * |: state   :|: "reg->valid" :|: "reg->dirty" :|
+ * | "invalid" | false          | false          |
+ * | "valid"   | true           | false          |
+ * | "dirty"   | true           | true           |
+ *
+ * The state "reg->valid == dirty && reg->valid == false" should be unreacheable.
+ * To achieve this, "reg->valid" and "reg->dirty" are not assigned directly
+ * outside of "mark_<state>()" functions.
+ */
+static void mark_invalid(struct reg *reg)
+{
+	assert(riscv_reg_impl_is_initialized(reg));
+	assert(reg->exist);
+	reg->valid = false;
+	reg->dirty = false;
+}
+
+static void mark_clean(struct reg *reg)
+{
+	assert(riscv_reg_impl_is_initialized(reg));
+	assert(reg->exist);
+	assert(riscv_reg_impl_get_target(reg)->state == TARGET_HALTED);
+	reg->valid = true;
+	reg->dirty = false;
+}
+
+static void mark_dirty(struct reg *reg)
+{
+	assert(riscv_reg_impl_is_initialized(reg));
+	assert(reg->exist);
+	assert(riscv_reg_impl_get_target(reg)->state == TARGET_HALTED);
+	reg->valid = true;
+	reg->dirty = true;
+}
+
+static int access_prelude(const struct reg *reg, bool write)
```

Comment:
Pedantic detail: Since the other function is called `epilogue`, this one can be called `prologue` (for naming symmetry).

- prelude vs. postlude
- prologue vs. epilogue

```suggestion
static int access_prologue(const struct reg *reg, bool write)
```

Reply:
Addressed. Thanks!